wled00/FX.cpp

```diff
@@ -1929,22 +1929,89 @@ static const char _data_FX_MODE_JUGGLE[] PROGMEM = "Juggle@!,Trail;;!;;sx=64,ix=
+  constexpr mathType sInt16Scale = 1.0f;
+  constexpr mathType maxAngle = M_TWOPI / 256.0;
+  constexpr mathType staticRotationScale = 1.0f;
+  constexpr mathType animatedRotationScale = M_TWOPI / double(0xFFFF);
+  constexpr float (*sinFunction)(float) = &sin_t;
+  constexpr float (*cosFunction)(float) = &cos_t;
+#endif
+  const bool isMatrix = strip.isMatrix;
+  const int cols = SEGMENT.virtualWidth();
+  const int rows = isMatrix ? SEGMENT.virtualHeight() : strip.getSegmentsNum();
+
+  const int inputShift = SEGMENT.speed;
+  const int inputSize = SEGMENT.intensity;
+  const int inputRotation = SEGMENT.custom1;
+  const bool inputAnimateShift = SEGMENT.check1;
+  const bool inputAnimateRotation = SEGMENT.check2;
+  const bool inputAssumeSquare = SEGMENT.check3;
+
+  const int paletteOffset = (!inputAnimateShift) ? (inputShift) : (((strip.now * ((inputShift >> 3) +1)) & 0xFFFF) >> 8);
+
+  mathType sinTheta;
+  mathType cosTheta;
+  if (rows <= 1) {
+    sinTheta = 0;
+    cosTheta = sInt16Scale;
+  } else if (cols <= 1) {
+    sinTheta = sInt16Scale;
+    cosTheta = 0;
+  } else {
+    const angleType theta = (!inputAnimateRotation) ? (inputRotation * maxAngle / staticRotationScale) : (((strip.now * ((inputRotation >> 4) +1)) & 0xFFFF) * animatedRotationScale);
+    sinTheta = sinFunction(theta);
+    cosTheta = cosFunction(theta);
+  }
+
+  const mathType maxX = std::max(1, cols-1);
+  const mathType maxY = std::max(1, rows-1);
+  const mathType maxXIn  =  inputAssumeSquare ? maxX : mathType(1);
+  const mathType maxYIn  =  inputAssumeSquare ? maxY : mathType(1);
+  const mathType maxXOut = !inputAssumeSquare ? maxX : mathType(1);
+  const mathType maxYOut = !inputAssumeSquare ? maxY : mathType(1);
+  const mathType centerX = sInt16Scale * maxXOut / mathType(2);
+  const mathType centerY = sInt16Scale * maxYOut / mathType(2);
+  const mathType scale   = std::abs(cosTheta) + (std::abs(sinTheta) * maxYOut / maxXOut);
+  const int yFrom = isMatrix ? 0 : strip.getCurrSegmentId();
+  const int yTo = isMatrix ? maxY : yFrom;
+  for (int y = yFrom; y <= yTo; ++y) {
+    const mathType ytSinTheta = mathType((wideMathType(sinTheta) * wideMathType(y * sInt16Scale - centerY * maxYIn))/wideMathType(maxYIn * scale));
+    for (int x = 0; x < cols; ++x) {
+      const mathType xtCosTheta = mathType((wideMathType(cosTheta) * wideMathType(x * sInt16Scale - centerX * maxXIn))/wideMathType(maxXIn * scale));
+      const mathType sourceX = xtCosTheta + ytSinTheta + centerX;
+      int colorIndex = (std::min(std::max(sourceX, mathType(0)), maxXOut * sInt16Scale) * 255) / (sInt16Scale * maxXOut);
+      if (inputSize <= 128) {
+        colorIndex = (colorIndex * inputSize) / 128;
+      } else {
+        // Linear function that maps colorIndex 128=>1, 256=>9
+        colorIndex = ((inputSize - 112) * colorIndex) / 16;
+      }
+      colorIndex += paletteOffset;
+      const uint32_t color = SEGMENT.color_wheel((uint8_t)colorIndex);
+      if (isMatrix) {
```

Comment:
use `SEGMENT.is2D()` instead of `strip.isMatrix`

Reply:
Hm. The UI won't let you configure a matrix that is only one wide or tall, so I guess `SEGMENT.is2D()` should always return the same as `strip.isMatrix`. However, using `strip.isMatrix` is a single boolean read, while `SEGMENT.is2D()` leads to quite some function calls and comparisons.  
I'm saying this just because I am wondering what the benefit of is2D is?

Additionally, you were talking about 1D segments. 1D segments would lead to `SEGMENT.is2D()` returning false. Which would then lead to my code switching from matrix mode to segmented mode, which leads to incorrect results.

Reply:
No, you can create a segment that is 1 pixel wide or 1 pixel high. But it is still on a matrix.

`is2D()` will return `true` if a segment has width and height > 1.

Reply:
Not sure if I made that clear enough: I tried your suggestion with 1D segments, and they were mostly black with is2D. With isMatrix they do show stuff. Although 1D segments never rotate, because a rotation doesn't make sense in 1D space.

The segment awareness only makes sense if we are not dealing with an actual 2D setup (if it has 1D segments). Otherwise why would you segment the matrix if you didn't want independent effects

Reply:
You got me thinking that rotating and having 1D segment would make a very nice effect.

You just need to adjust code to operate in "square" mode regardless.